cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,869 @@
+  /**
+   * @brief Functor to compute distance from a given row to the edge
+   * of the window.
+   *
+   * @tparam Grouping Object defining how the orderby column is
+   * grouped.
+   * @tparam OrderbyT Type of elements in the orderby columns.
+   * @tparam DeltaT Type of the elements in the scalar delta (returned
+   * by scalar.data()).
+   * @param groups The grouping object.
+   * @param row_delta Pointer to row delta on device.
+   * @param begin Iterator to begin of orderby column on device.
+   * @param end Iterator to end of orderby column on device.
+   *
+   * @note Let `x` be the value of the current row and `delta` the provided
+   * row delta then for bounded windows the endpoints are computed as follows.
+   *
+   *           | ASCENDING | DESCENDING
+   * ----------+-----------+-----------
+   * PRECEDING | x - delta | x + delta
+   * FOLLOWING | x + delta | x - delta
+   *
+   * See `saturating_sub` and `saturating_add` for details of the implementation of
+   * saturating addition/subtraction.
+   */
+  template <typename Grouping, typename OrderbyT, typename DeltaT>
+  struct distance_kernel {
+    Grouping groups;  ///< Group information to determine bounds on current row's window
+    static_assert(cuda::std::is_same_v<Grouping, ungrouped> ||
+                    cuda::std::is_same_v<Grouping, grouped> ||
+                    cuda::std::is_same_v<Grouping, ungrouped_with_nulls> ||
+                    cuda::std::is_same_v<Grouping, grouped_with_nulls>,
+                  "Invalid grouping descriptor");
+    DeltaT const* row_delta;  ///< Delta from current row that defines the interval endpoint. This
+                              ///< pointer is null for UNBOUNDED and CURRENT_ROW windows.
+    column_device_view::const_iterator<OrderbyT>
+      begin;                                           ///< Iterator to beginning of orderby column
+    column_device_view::const_iterator<OrderbyT> end;  ///< Iterator to end of orderby column
+
+    /**
+     * @brief Compute the offset to the end of the window.
+     *
+     * @param i The current row index.
+     * @return Offset to the current row's window endpoint.
+     */
+    [[nodiscard]] __device__ size_type operator()(size_type i) const
+    {
+      using Comp      = comparator_t<OrderbyT, WindowTag, Order>;
+      auto const info = groups.row_info(i);
+      auto const [null_count, group_start, group_end, null_start, null_end, start, end] =
+        groups.row_info(i);
+      if constexpr (Direction == direction::PRECEDING) {
+        if constexpr (WindowTag == window_tag::UNBOUNDED) { return i - group_start + 1; }
+        // TODO: If the window is BOUNDED_OPEN, what does it mean for a row to fall in the null
+        // group? Not that important because only spark allows nulls in the orderby column, and it
+        // doesn't have BOUNDED_OPEN windows.
+        if (Grouping::has_nulls && i >= null_start && i < null_end) { return i - null_start + 1; }
```

Comment:
Microopt, statically pruning this branch for columns without nulls (in the previous code, this branch was always checked).